src/liboslcomp/oslcomp.cpp

```diff
@@ -1073,83 +1076,69 @@ OSLCompilerImpl::track_variable_lifetimes (const OpcodeVec &code,
     BOOST_FOREACH (Symbol *s, allsyms)
         s->clear_rw ();
 
-    static ustring op_for("for");
-    static ustring op_while("while");
-    static ustring op_dowhile("dowhile");
+    // Keep track of the nested loops we're inside. We track them by pairs
+    // of begin/end instruction numbers for that loop body, including
+    // conditional evaluation (skip the initialization). Note that the end
+    // is inclusive. We use this vector of ranges as a stack.
+    typedef std::pair<int,int> intpair;
+    std::vector<intpair> loop_bounds;
 
     // For each op, mark its arguments as being used at that op
     int opnum = 0;
     BOOST_FOREACH (const Opcode &op, code) {
-        // Some work to do for each argument to the op...
-        for (int a = 0;  a < op.nargs();  ++a) {
-            SymbolPtr s = opargs[op.firstarg()+a];
-            ASSERT (s->dealias() == s);
-            // s = s->dealias();   // Make sure it's de-aliased
-
-            // Mark that it's read and/or written for this op
-            s->mark_rw (opnum, op.argread(a), op.argwrite(a));
-        }
-
-        // If this is a loop op, we need to mark its control variable
-        // (the only arg) as used for the duration of the loop!
-        if (op.opname() == op_for ||
-            op.opname() == op_while ||
-            op.opname() == op_dowhile) {
+        if (op.opname() == op_for || op.opname() == op_while ||
+                op.opname() == op_dowhile) {
+            // If this is a loop op, we need to mark its control variable
+            // (the only arg) as used for the duration of the loop!
             ASSERT (op.nargs() == 1);  // loops should have just one arg
             SymbolPtr s = opargs[op.firstarg()];
+            int loopcond = op.jump (0); // after initialization, before test
+            int loopend = op.farthest_jump() - 1;   // inclusive end
             s->mark_rw (opnum+1, true, true);
-            s->mark_rw (op.farthest_jump()-1, true, true);
+            s->mark_rw (loopend, true, true);
+            // Also push the loop bounds for this loop
+            loop_bounds.push_back (std::pair<int,int>(loopcond, loopend));
```

Comment:
Yes, I'll do that. 

make_shared is C++11, make_unique is C++14, I'd forgotten that make_pair was old school.

Some days, still working in C++98 makes me feel like I may as well be using Fortran 77.